// @bagatell/core/src/index.js
// Fundamental classes
export { App } from './classes/App.js'
export { Direction } from './classes/Direction.js'
export { FrameworkApp } from './classes/FrameworkApp.js'
export { Page } from './classes/Page.js'

// Functions
export { createFrameworkApp } from './functions/create-framework-app.js'

// Components
import { createComponentCreator } from './functions/create-component-creator.js'

import {ButtonComponent} from './components/ButtonComponent.js'
export const Button = createComponentCreator(ButtonComponent)

import {EnterTextComponent} from './components/EnterTextComponent.js'
export const EnterText = createComponentCreator(EnterTextComponent)

import {EnterNumberComponent} from './components/EnterNumberComponent.js'
export const EnterNumber = createComponentCreator(EnterNumberComponent)

import {ImageComponent} from './components/ImageComponent.js'
export const Image = createComponentCreator(ImageComponent)

import {SliderComponent} from './components/SliderComponent.js'
export const Slider = createComponentCreator(SliderComponent)

import {TextComponent} from './components/TextComponent.js'
export const Text = createComponentCreator(TextComponent)

import {TimerComponent} from './components/TimerComponent.js'
export const Timer = createComponentCreator(TimerComponent)

import {ColsComponent} from './components/ColsComponent.js'
export const Cols = createComponentCreator(ColsComponent)

import {LayersComponent} from './components/LayersComponent.js'
export const Layers = createComponentCreator(LayersComponent)

import {RowsComponent} from './components/RowsComponent.js'
export const Rows = createComponentCreator(RowsComponent)

import {SpaceComponent} from './components/SpaceComponent.js'
export const Space = createComponentCreator(SpaceComponent)
